### DISCUSSION TITLE
Detached entity passed to persist

### DISCUSSION THREAD

**Discussion by utnFBlasco** [dds-jv/foro#315] [Consultas de persistencia]
Buenas tenemos un error en la independencia de los tests cuando implementamos la base de datos, nos tira el error 'detached entity passed to persist', googleamos el error y hay muchas respuestas pero no nos sirvió ninguna. Una que resaltaba era la que se estaba creando una fila con un id ya generado, pero debugeando chequeamos la base de datos y se hace bien el rollback sobre el test anterior, por lo tanto no tendría sentido ya que las tablas están vacías. El error como se puede ver en la imagen, se produce en la línea getListaPreguntas(busca en la bd las preguntas) pero eso no es parte del test y ya esta testeado en otros test. Además el error hace referencia a la entidad dato de contacto(es una tabla que se relaciona con la de usuarios) y no a la entidad de las preguntas que tampoco es la entidad que queremos testear sino que a lo que apunta el test son las publicaciones.

![image](https://user-images.githubusercontent.com/82328433/135473012-aa8d2114-2367-4a69-8210-a08fd055a467.png)

Si alguno lo resolvió y nos puede dar una mano buenísimo, Saludos! Abajo paso los dos tests que ejecutados juntos rompen.

```java
@Test
  public void seLeEnviaRecomendacionSemanales() {

  PublicacionIntencionAdopcion publicacionIntension = generarPublicacionIntencionAdopcionSimple(
      datosPublicacion1,
      datosHarry
  );
  PublicacionAdopcionMascota publicacionAdopcionNoCoincidente = generarPublicacionAdopcionSimple(
      datosPublicacion3,
      datosPepe
  );
  PublicacionAdopcionMascota publicacionAdopcionCoincidente = generarPublicacionAdopcionSimple(
      datosPublicacion1,
      datosPepe
  );
  patitas.generarPublicacionIntencionAdopcion(publicacionIntension);
  patitas.generarPublicacionParaAdopcion(publicacionAdopcionCoincidente);
  patitas.generarPublicacionParaAdopcion(publicacionAdopcionNoCoincidente);
  patitas.enviarRecomendaciones();

  Mockito.verify(notificacionFalsa, Mockito.times(2)).enviarNotificacion(Mockito.any(),Mockito.anyString(),Mockito.anyString());
}


@Test
  public void filtraLasPublicacionesSegunPreferencias() {
    PublicacionIntencionAdopcion publicacionIntension = generarPublicacionIntencionAdopcionSimple(
        datosPublicacion1,
        datosHarry
    );
    PublicacionAdopcionMascota publicacionAdopcionNoCoincidente = generarPublicacionAdopcionSimple(
        datosPublicacion3,
        datosPepe
    );
    PublicacionAdopcionMascota publicacionAdopcionCoincidente = generarPublicacionAdopcionSimple(
        datosPublicacion1,
        datosPepe
    );
    patitas.generarPublicacionIntencionAdopcion(publicacionIntension);
    patitas.generarPublicacionParaAdopcion(publicacionAdopcionCoincidente);
    patitas.generarPublicacionParaAdopcion(publicacionAdopcionNoCoincidente);
    List<PublicacionAdopcionMascota> listaPublicacionesFiltradas = patitas.filtrarPublicacionesInteresadosAdopcion(publicacionIntension);
    assertTrue(listaPublicacionesFiltradas.contains(publicacionAdopcionCoincidente));
  }
}
```

**Comment by gastonprieto** [comment]
buenas @utnFBlasco

> El error como se puede ver en la imagen, se produce en la línea getListaPreguntas(busca en la bd las preguntas) pero eso no es parte del test

por lo que veo del stack trace ese metodo (`getListaPreguntas`) se llama porque llamas al `generarPublicacionParaAdopcion`, revisate que en el stacktrace todas las llamadas.

>Además el error hace referencia a la entidad dato de contacto(es una tabla que se relaciona con la de usuarios) y no a la entidad de las preguntas que tampoco es la entidad que queremos testear sino que a lo que apunta el test son las publicaciones.

Puede ser que datosPepe sea de ese tipo? y de alguna forma este involucrada en la busqueda?

En general hibernate bufferea operaciones, por ejemplo un update, tambien podria ser una creacion, hasta que es necesaria. Por ejemplo puede que este buffereando el insert, pero como luego tiras un select, dice.. momento, tengo un insert pendiente, y cuando lo va a tirar dice, momento, este objeto no lo tengo marcado como persist ni tampoco es que le pusieron el cascade, entonces te rompe ahi.

**Reply by utnFBlasco** [reply]
datosPepe es de ese tipo y esta involucrada, pero igualmente la tenemos marcada como entidad DatoDeContacto con sus anotaciones de cascada en la publicación. Gracias por contestar.

**Reply by gastonprieto** [reply]
@utnFBlasco puede ser que la cascada no este en usuario? fijate que te dice el detach en usuario.datosDeContacto
